src/providers/postgres/qgspostgresfeatureiterator.cpp

```diff
@@ -111,12 +112,34 @@ QgsPostgresFeatureIterator::QgsPostgresFeatureIterator( QgsPostgresFeatureSource
     }
   }
 
-  bool success = declareCursor( whereClause, limitAtProvider ? mRequest.limit() : -1, false );
+  QStringList orderByParts;
+
+  mOrderByCompiled = true;
+
+  Q_FOREACH ( const QgsFeatureRequest::OrderByClause& clause, request.orderBys() )
+  {
+    QgsPostgresExpressionCompiler compiler = QgsPostgresExpressionCompiler( source );
+    QgsExpression expression = clause.expression();
+    if ( compiler.compile( &expression ) == QgsSqlExpressionCompiler::Complete )
+    {
+      QString part;
+      part = compiler.result();
+      part += clause.ascending() ? " ASC" : " DESC";
+      part += clause.nullsFirst() ? " NULLS FIRST" : " NULLS LAST";
+      orderByParts << part;
+    }
+    else
+    {
+      mOrderByCompiled = false;
```

Comment:
Not sure.

Partial compilation for sure won't help.

Having the first of two clauses pass may help to presort and reduce the time the local sort fallback requires (but the expensive expression evaluation will still need to take place)

I think it should stop at the first failing clause. This way it will only use partially compiled important clauses and not pre-sort on unimportant ones.